src/mem_table.rs

```diff
@@ -31,10 +31,16 @@ pub(crate) struct ImmutableWal {
     table: Arc<KVTable>,
 }
 
-type MemTableRange<'a> = Range<'a, Bytes, (Bound<Bytes>, Bound<Bytes>), Bytes, ValueDeletable>;
+type MemTableRange<'a> = Range<'a, Bytes, (Bound<Bytes>, Bound<Bytes>), Bytes, MemValue>;
 
 pub struct MemTableIterator<'a>(MemTableRange<'a>);
 
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct MemValue {
```

Comment:
Ah, I see why we needed `val.value` earlier. Is there a more elegant way to refactor this that doesn't require a both `ValueDeletable` and `MemValue`? We currently have `KeyValue`, `KeyValueDeletable`, `ValueDeletable`, and now `MemValue`. This seems excessive.

Reply:
Yeah, that's a good point. I'll look into it, I remember thinking exactly this and they all actually served different purposes (different combos of what's passed around) but I'll see if there's anything that can be done.

Reply:
I think having this type makes sense (a value/tombstone + attributes). We might want to rename it (to I'm not sure what) since `MemValue` didn't really make sense to me. Maybe `ValueDeletableWithAttributes`.

A type for a key+value/tombstone+attributes (`KeyValueDeletable`) is required for iteration. We could rename that to keep it consistent with this type (e.g. `KeyValueDeletableWithAttributes`).

I actually think there isn't a use case for `KeyValue` outside of tests. Maybe we just clean that up?

Basically I think outside of `KeyValue` there are good reasons for having the existing types. We just maybe need to rename them, but other than `MemValue` I don't think we should do that in this PR.

Reply:
I think my brain can handle `ValueDeletableWithAttributes`. But isn't deletable an attribute now? It's a flag right?

Reply:
> But isn't deletable an attribute now?

It's only an attribute in the encoded format, it's a lot less error-prone to have a statically typed enum for Value(val), Tombstone than to require people to check the attributes to know whether or not there's a value (which I assume would be optional). I can make that change to the memory model though if we think that makes things simpler.

FYI `MemValue` was to indicate it's only used in the memtable. I can make the change Rohan requested though, it's just a little verbose 😆

Reply:
i suppose `ValueWithAttributes` might be better when compares with `ValueDeletableWithAttributes ` if deletion is already considered as an attribute/flag.

Reply:
sorry after reading the rest part, i suspect we may better to encode the attributes in the `Key` instead of `Value` struct if the skiplist does not hold multiple values for a single key.

related comment: https://github.com/slatedb/slatedb/pull/265#discussion_r1807685961

this is the encoding in leveldb's memtable: https://github.com/google/leveldb/blob/23e35d792b9154f922b8b575b12596a4d8664c65/db/memtable.cc#L76 (it encoded both the key & value for the skiplist though..)

Reply:
There are many places in the code that leverage the enum `Value(bytes), Tombstone` so I'm going to leave this as is for this PR to minimize the code changes. we can change it in a follow-up if we want.

I named it `ValueWithAttributes` for verbosity sake, but it contains a `ValueDeletable` as a field.